test/test_helper.rb

```diff
@@ -60,17 +60,26 @@ class ActiveSupport::TestCase
   fixtures :all
 
   before do
+    @before_threads = Thread.list
     Rails.cache.clear
     create_default_stubs
   end
 
-  after { sleep 0.1 while extra_threads.present? }
+  after { fail_if_dangling_threads }
+
+  def fail_if_dangling_threads
+    max_threads = ENV['CI'] ? 0 : 1
```

Comment:
that means each test can add 1 more thread like `3,4,5,6,7` vs being capped at 4 before